service/storage_proxy.cc

```diff
@@ -4593,6 +4595,36 @@ future<bool> storage_proxy::cas(schema_ptr schema, shared_ptr<cas_request> reque
     co_return condition_met;
 }
 
+bool storage_proxy::is_likely_to_respond_in_time(const gms::inet_address& ep, clock_type::time_point timeout) {
+    if (fbu::is_me(ep)) {
+        return true;
+    }
+    if (auto last_seen_ep = last_seen(ep)) {
+        auto now = clock_type::now();
+        auto time_without_response = last_seen_ep->last_sent - last_seen_ep->last_responded;
+        auto time_left = timeout - now;
+        auto last_attempt = now - last_seen_ep->last_sent;
+        // If the node has never responded yet or there was no communication attempt for a long while,
+        // the node should be probed in case it becomes responsive
+        if (last_seen_ep->last_responded == clock_type::time_point::min() || last_attempt > time_left) {
+            return true;
+        }
+        if (time_without_response > time_left) {
+            static constexpr int granularity = 10'000;
+            static thread_local std::uniform_int_distribution dice(0, granularity);
+            // There's always a small chance of probing the replica in case it came back to life.
+            int fail_chance = std::min<long>((time_without_response.count() - time_left.count()) * granularity / time_left.count(), granularity - 1);
```

Comment:
Are we certain that time_left is nonzero and nonnegative?

Reply:
Hm, as long as timeout > now, which sounds reasonable, but perhaps it's also not guaranteed. That can be short-circuited with a simple check - if we're past timeout already, the node can be safely qualified as dead - as will other nodes, and the query would deservedly fail.